File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode prev=null;                       // Previous Node ka val store krney k liye; shuru mei null hoga
    boolean ans=true;                         // Ans shuru mei shi hai yni valid hai

    // Inorder Traversal krna hai
    public void inorder(TreeNode root){
        // Agar Root hi null hai; yni Tree hi nhi hai
        if(root==null){
            return;
        }

        // Left Child ko mauka dedo
        inorder(root.left);

        // Khud ko mauka dedo
        if(prev==null){
            // Yni ki first node hai
            prev=root;                        // Root bna loo
        }
        else{
            if(root.val<=prev.val){
                // Agar Root se Prev bda hai; tab toh BST hold nhi hoga
                ans=false;
            }
            prev=root;
        }

        // Right Child ko mauka Dedo
        inorder(root.right);
    }

    public boolean isValidBST(TreeNode root) {
        // Recursive Function Call karo
        inorder(root);

        return ans;
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
TreeNode* prev=nullptr;                       // Previous Node ka val store krney k liye ; shuru mei 0 hoga
bool ans=true;                               // ans khali hai and shuru mei shi hai yni valid hai

// Inoreder trversal krna hai 
void inorder(TreeNode*root){
    // Agar root hi null hai; ynhi tree hi nhi hai
    if(root==nullptr){
        return;
    }

    // Left child ko mauka dedo 
    inorder(root->left);

    // Khud ko mauka dedo 
    if(prev==nullptr){
        // yni ki first node hai 
        prev=root;                              // root bna loo 
    }
    else{
        if(root->val<=prev->val){
            // Agar root se prev bda hai; tab toh BST hold nhi hoga 
            ans=false;
        }
        prev=root;
    }

    // Right Child Ko mauka Dedo
    inorder(root->right);
}

    bool isValidBST(TreeNode* root) {
        // Recursive Function Call karo 
        inorder(root);

        return ans;  
    }
};